### ISSUE TITLE
`min` and `max` definitions in `Arduino.h` cause compilation errors

### ISSUE THREAD

**Issue by pdcook** [h2zero/n-able-Arduino#12]
Any project which uses `std::unordered_map` fails to compile. The error message is extraordinarily long, but starts as:

```
rom /home/pcook/.arduino15/packages/h2zero/tools/gcc-arm-none-eabi/9.3.1-1/arm-none-eabi/include/c++/9.3.1/bits/hashtable.h:35,
                 from /home/pcook/.arduino15/packages/h2zero/tools/gcc-arm-none-eabi/9.3.1-1/arm-none-eabi/include/c++/9.3.1/unordered_map:46,
/home/pcook/.arduino15/packages/h2zero/tools/gcc-arm-none-eabi/9.3.1-1/arm-none-eabi/include/c++/9.3.1/limits:1818:7: error: 'constexpr' static data member 'max' must have an initializer
 1818 |       max() _GLIBCXX_USE_NOEXCEPT { return __LDBL_MAX__; }
      |       ^~~
/home/pcook/.arduino15/packages/h2zero/tools/gcc-arm-none-eabi/9.3.1-1/arm-none-eabi/include/c++/9.3.1/limits:1818:7: error: expected ';' at end of member declaration
 1818 |       max() _GLIBCXX_USE_NOEXCEPT { return __LDBL_MAX__; }
      |       ^~~
      |          ;
```

I hope this is just something I'm doing wrong. An identical project using the non-n-able core compiles without issue.

**Comment by h2zero** [comment]
Interesting, thanks for the report. I'll look into this asap.

**Comment by h2zero** [comment]
Looks like there is a redefinition of min and max in the "drivers" library. Is this a private lib?

**Comment by pdcook** [comment]
Yes it is, but I just did a quick `cat * | grep min` and the same for `max` and I haven't redefined them, or even used them as a variable name. At worst they are _part_ of a variable name, or in a comment. Very strange, and stranger still that this compiles fine on the default Arduino core.

**Comment by pdcook** [comment]
Oh boy, I think I've found it: https://arduinojson.org/v6/error/macro-min-passed-3-arguments-but-takes-just-2/

Looks like `Arduino.h` and `std::min` conflict. I'll see if I can fix this by changing the order of the `#includes`.

**Comment by pdcook** [comment]
I have confirmed that the issue is indeed in the `Arduino.h` definition of `min` and `max.` Commenting those lines out results in error-free compilation.

**Comment by h2zero** [comment]
Awesome, any thoughts about a permanent solution? Not sure if they did anything in the samd core to deal with this.

**Comment by pdcook** [comment]
@h2zero It's unclear to me how exactly they fixed it in the samd core. Something to do with the underlying arduino-api. But here a permanent (albeit a bit sloppy) fix would just be to use the same

```
#ifdef min
#undef min
#endif
```

that the samd core uses for `abs`. However, for whatever reason `#ifdef min` is `false`! As is `#ifdef` of any substring of `min`! So this doesn't work. Simply commenting out the definitions of `min` and `max` does, but that can't be a reasonable solution.

**Comment by h2zero** [comment]
@pdcook Sorry for the delay on this, I have created PR #15 which implements the fix from the arduinocore-api. If you could test your code with that PR and let me know if it works for you I would greatly appreciate it.

**Comment by h2zero** [comment]
Closing as stale/fixed
